src/Reflection/ReflectionClass.php

```diff
@@ -1056,7 +1060,28 @@ public function getParentClass(): ReflectionClass|null
      */
     public function getParentClassNames(): array
     {
-        return array_map(static fn (self $parentClass): string => $parentClass->getName(), array_slice(array_reverse($this->getInheritanceClassHierarchy()), 1));
+        if ($this->cachedParentClassNames === null) {
+            $parentClassNames = [];
+
+            $parentClass = $this->getParentClass();
+            while ($parentClass !== null) {
+                $parentClassName = $parentClass->getName();
+
+                if (
+                    $this->name === $parentClassName
+                    || in_array($parentClassName, $parentClassNames, true)
+                ) {
+                    throw CircularReference::fromClassName($parentClassName);
+                }
+
+                $parentClassNames[] = $parentClassName;
+                $parentClass        = $parentClass->getParentClass();
+            }
+
+            $this->cachedParentClassNames = $parentClassNames;
+        }
```

Comment:
Can we implement this check without having a mutable `$this->cachedParentClassNames`? Thinking of a by-ref parameter in a `private` method :thinking:

Reply:
Is is such a problem? `ReflectionClass` already have some more caches...